From e2ea87a1d9e68bd984afccd9ed245b5ad3b31f6c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marcin=20Wcis=C5=82o?= <marcin.wcislo@conclusive.pl>
Date: Fri, 22 Aug 2025 19:33:41 +0200
Subject: [PATCH] drm/vmwgfx: Fix invalid reads in fence signaled events
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

jira VULN-8241
cve CVE-2024-36960
commit-author Zack Rusin <zack.rusin@broadcom.com>
commit a37ef7613c00f2d72c8fc08bd83fb6cc76926c8c

Correctly set the length of the drm_event to the size of the structure
that's actually used.

The length of the drm_event was set to the parent structure instead of
to the drm_vmw_event_fence which is supposed to be read. drm_read
uses the length parameter to copy the event to the user space thus
resuling in oob reads.

	Signed-off-by: Zack Rusin <zack.rusin@broadcom.com>
Fixes: 8b7de6aa8468 ("vmwgfx: Rework fence event action")
	Reported-by: zdi-disclosures@trendmicro.com # ZDI-CAN-23566
	Cc: David Airlie <airlied@gmail.com>
CC: Daniel Vetter <daniel@ffwll.ch>
	Cc: Zack Rusin <zack.rusin@broadcom.com>
	Cc: Broadcom internal kernel review list <bcm-kernel-feedback-list@broadcom.com>
	Cc: dri-devel@lists.freedesktop.org
	Cc: linux-kernel@vger.kernel.org
	Cc: <stable@vger.kernel.org> # v3.4+
	Reviewed-by: Maaz Mombasawala <maaz.mombasawala@broadcom.com>
	Reviewed-by: Martin Krastev <martin.krastev@broadcom.com>
Link: https://patchwork.freedesktop.org/patch/msgid/20240425192748.1761522-1-zack.rusin@broadcom.com
(cherry picked from commit a37ef7613c00f2d72c8fc08bd83fb6cc76926c8c)
	Signed-off-by: Marcin Wcisło <marcin.wcislo@conclusive.pl>
---
 drivers/gpu/drm/vmwgfx/vmwgfx_fence.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/vmwgfx/vmwgfx_fence.c b/drivers/gpu/drm/vmwgfx/vmwgfx_fence.c
index 2a0cda324703..5efc6a766f64 100644
--- a/drivers/gpu/drm/vmwgfx/vmwgfx_fence.c
+++ b/drivers/gpu/drm/vmwgfx/vmwgfx_fence.c
@@ -991,7 +991,7 @@ static int vmw_event_fence_action_create(struct drm_file *file_priv,
 	}
 
 	event->event.base.type = DRM_VMW_EVENT_FENCE_SIGNALED;
-	event->event.base.length = sizeof(*event);
+	event->event.base.length = sizeof(event->event);
 	event->event.user_data = user_data;
 
 	ret = drm_event_reserve_init(dev, file_priv, &event->base, &event->event.base);
-- 
2.47.3

